src/hotspot/share/gc/z/zBarrierSet.inline.hpp

```diff
@@ -479,24 +479,54 @@ inline void ZBarrierSet::AccessBarrier<decorators, BarrierSetT>::clone_in_heap(o
   ZIterator::oop_iterate(dst, &cl_sg);
 }
 
+static inline void copy_primitive_payload(const void* src, const void* dst, const size_t payload_size_bytes, size_t& copied_bytes) {
+  if (payload_size_bytes == 0) {
+    return;
+  }
+  void* src_payload = (void*)(address(src) + copied_bytes);
+  void* dst_payload = (void*)(address(dst) + copied_bytes);
+  Copy::copy_value_content(src_payload, dst_payload, payload_size_bytes);
+  copied_bytes += payload_size_bytes;
+}
+
 template <DecoratorSet decorators, typename BarrierSetT>
 inline void ZBarrierSet::AccessBarrier<decorators, BarrierSetT>::value_copy_in_heap(void* src, void* dst, InlineKlass* md, LayoutKind lk) {
   if (md->contains_oops()) {
-    // src/dst aren't oops, need offset to adjust oop map offset
-    const address src_oop_addr_offset = ((address) src) - md->first_field_offset();
+    // Iterate over each oop map, performing:
+    //   1) possibly raw copy for any primitive payload before each map
+    //   2) load and store barrier for each oop
+    //   3) possibly raw copy for any primitive payload trailer
+    assert(lk == NON_ATOMIC_FLAT || lk == PAYLOAD, "Cannot support layout other than NON_ATOMIC_FLAT"); // Note: PAYLOAD is incorrect, resolve when transistioned to new flattening
```

Comment:
Why is PAYLOAD incorrect here?
And what about ATOMIC_FLAT? Even with ZGC, it can contain an oop.

Reply:
You are correct there is a use single case under the `md->contains_oops()` condition with Z that is valid, a single oop. Will adjust